### PR TITLE
xds-failover: prevent sending initial_resource_versions when moving from fallback to primary

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: xds-failover: prevent sending initial_resource_versions when moving from fallback to primary
Additional Description:
The [initial_resource_versions](https://github.com/envoyproxy/envoy/blob/3283bedf6700276477e681294b210ee3a01cde56/api/envoy/service/discovery/v3/discovery.proto#L236) field is used in delta-xDS s.t. Envoy supplies the versions of the resources it is aware of, and the server doesn't need to send these resources if they haven't been modified.
This PR updates Envoy to avoid populating that field when connecting to the primary source after being connected to the failover source.
Risk Level: low
Testing: Added integration tests.
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
@htuch unfortunately this PR required some small updates to many tests. Please take a pass on this, and LMK if this should be broken even further.
/assign @htuch

**Comment by htuch** [comment]
Is the intuition that the primary version might be different or more up-to-date? Shouldn't this be captured in the version itself?

**Comment by adisuissa** [comment]
> Is the intuition that the primary version might be different or more up-to-date? Shouldn't this be captured in the version itself?

Theoretically speaking, the version needs to be determined by the resource contents (e.g., hash of the contents).
The assumption is that the primary source always has same or better view of the resources.
The request is to "reset" the resources when moving from the failover to the primary, so no lingering resources will be kept in the Envoy.

**Comment by adisuissa** [comment]
FWIW, the docs describing the xDS-Failover feature are in #35746 (which is currently in draft mode until this is merged).

File: envoy/config/grpc_mux.h
Comment (adisuissa) [reply envoy/config/grpc_mux.h]: Oops. fixed.

File: envoy/config/grpc_mux.h
Comment (adisuissa) [reply envoy/config/grpc_mux.h]: Done.

File: envoy/config/grpc_mux.h
Comment (adisuissa) [reply envoy/config/grpc_mux.h]: You are correct.
I've renamed the parameter's name.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:228]: I thought we always sent initial resource version when connecting to the failover?

**Review by htuch — COMMENTED** [review]
LGTM modulo comment (and assuming perfect test coverage ;-)

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:228]: At the moment the failover is only being attempted if no resources were fetched from the primary (thus no initial versions should exist).

**Comment by adisuissa** [comment]
> LGTM modulo comment (and assuming perfect test coverage ;-)

:) Trying to rerun the flaky test, to get the coverage report.

**Comment by adisuissa** [comment]
> LGTM modulo comment (and assuming perfect test coverage ;-)

coverage LGTM. We'll need 2 stamps to merge.

**Comment by nezdolik** [comment]
I will take a look in couple of hours.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:228]: Maybe a good candidate for an ASSERT?

Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:228]: To be clear the code line mentioned above handles a case where
1. Envoy connected to failover (and fetches some resources)
2. Envoy is disconnected from the failover and attempts to connect to the primary - no initial_resource_versions will be sent.
3. Envoy never connects to the primary and attempts to connect to the failover - will send the initial_resource_versions that were fetched in (1).

RE the assert - there's an assertion that no failover stream is established when `ever_connected_to_primary_` is true.

**Review by nezdolik — APPROVED** [review]
Looks good!

**Review by htuch — APPROVED** [review]

**Pull request merged**
